### PR TITLE
 IGNITE-17709 SQL: add support for table hints.

### PR DESCRIPTION
**Pull request by xtern**

### CODE REVIEW

**Comment by AMashenkov** [comment]
What is the behaviour in case of typo in column name? E.g. " use_index(unknown_col)". 
Should we test this as well?

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/PlannerTest.java [hunk 1]
```diff
@@ -860,6 +866,81 @@ public void correctPlanningWithOrToUnion() throws Exception {
...
+
+            @Override public IgniteDistribution distribution() {
+                return IgniteDistributions.affinity(0, "ignored", "ignored");
+            }
+        };
+
+        IgniteSchema publicSchema = new IgniteSchema("PUBLIC");
+
+        publicSchema.addTable("PERSON", person);
+        publicSchema.addTable("COMPANY", company);
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/PlannerTest.java:898]:
```suggestion
        IgniteSchema publicSchema = createSchema(
                createTable("PERSON", IgniteDistributions.affinity(0, "ignored", "ignored"),
                        "PK", Integer.class, "ORG_ID", Integer.class
                ),
                createTable("COMPANY", IgniteDistributions.affinity(0, "ignored", "ignored"),
                        "PK", Integer.class, "ID", Integer.class
                )
        );
```

File: modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/PlannerTest.java [hunk 2]
```diff
@@ -860,6 +866,81 @@ public void correctPlanningWithOrToUnion() throws Exception {
...
+                                .build())
+                        .build())
+                .query(sql)
+                .build();
+
+        IgniteRel plan = physicalPlan(sql, ctx);
+
+        Map<String, Integer> hintsMap = new HashMap<>();
+
+        plan.accept(new RelShuttleImpl() {
```
Comment (korlov42) [inline modules/sql-engine/src/test/java/org/apache/ignite/internal/sql/engine/planner/PlannerTest.java:926]: please take a look at `org.apache.ignite.internal.sql.engine.planner.AbstractPlannerTest#assertPlan`. You may find it more convenient

**Review by korlov42 — COMMENTED** [review]
I think it worth to provide the proper implementation of `org.apache.calcite.rel.hint.Hintable#withHints` as well. Otherwise there is a chance that hints will be lost after applying some transformations

**Comment by xtern** [comment]
> I think it worth to provide the proper implementation of `org.apache.calcite.rel.hint.Hintable#withHints` as well. Otherwise there is a chance that hints will be lost after applying some transformations

Done, thanks

**Pull request merged**